src/DesktopIntegration.vala

```diff
@@ -107,12 +107,55 @@ public class Gala.DesktopIntegration : GLib.Object {
         foreach (unowned var app in apps) {
             foreach (weak Meta.Window window in app.get_windows ()) {
                 if (window.get_id () == uid) {
-                    window.get_workspace ().activate_with_focus (window, wm.get_display ().get_current_time ());
+                    if (window.has_focus ()) {
+                        notify_already_focused (window);
+                    } else {
+                        window.get_workspace ().activate_with_focus (window, wm.get_display ().get_current_time ());
+                    }
                 }
             }
         }
     }
 
+    private bool notifying = false;
+    private void notify_already_focused (Meta.Window window) {
+        if (notifying) {
+            return;
+        }
+
+        notifying = true;
+
+        wm.get_display ().get_sound_player ().play_from_theme ("bell", _("Window has already focus"), null);
+
+        if (window.get_maximized () == BOTH) {
+            notifying = false;
+            return;
+        }
+
+        var rect = window.get_frame_rect ();
+
+        int n = 0;
+        Timeout.add (10, () => {
+            int diff = 15;
+            if (n % 2 == 0) {
+                diff = -15;
+            }
+
+            window.move_frame (false, rect.x + diff, rect.y);
+
+            n++;
+
+            var should_continue = n <= 10;
+
+            if (!should_continue) {
+                notifying = false;
+                window.move_frame (false, rect.x, rect.y);
+            }
+
+            return should_continue;
+        });
+    }
```

Comment:
It feels to me that we should use Mutter integrated way to handle animations (which relies on frames instead of GLib time)

There is a similar utility in GNOME Shell for instance https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/misc/animationUtils.js#L80

Reply:
FYI this is how the "ease" function is defined https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/ui/environment.js#L92

Reply:
Hmm I get some weird flickering when doing this:

[Kooha-2024-08-07-15-39-04.webm](https://github.com/user-attachments/assets/12553b33-4512-4899-a096-b100ef1aeeb9)

Any idea why?

Code:
```vala
        var transition = new Clutter.KeyframeTransition ("translation-x") {
            repeat_count = 5,
            duration = 200,
            remove_on_complete = true
        };
        transition.set_from_value (0);
        transition.set_to_value (0);
        transition.set_values ( { -20, 20 });

        transition.stopped.connect (() => {
            notifying = false;
            wm.get_display ().enable_unredirect ();
        });

        wm.get_display ().disable_unredirect ();

        ((Meta.WindowActor) window.get_compositor_private ()).add_transition ("notify-already-focused", transition);
```

Reply:
You forgot to add key frames:

`transition.set_key_frames ( { 0.5, -0.5 } );`

Reply:
@lenemter thank you that fixed it!